deadline_/notifications/handlers.py

```diff
@@ -0,0 +1,261 @@
+
+    @staticmethod
+    def receive_message(msg: str) -> bool:
+        """
+        Processes the message and sends it to the handler
+        :return: a boolean indicating if hte message was successfully processed
+        """
+        try:
+            notif_id = int(msg)
+            notification = NotificationsHandler.fetch_notification(notif_id)
+
+            asyncio.ensure_future(NotificationsHandler.send_notification(notification))
+        except (NotificationAlreadyRead, Notification.DoesNotExist) as e:
+            print(f'DEBUG - Notification was either read or with an invalid ID - {e}')
+        except OfflineRecipientError as e:
+            print(f'Recipient was not eligible to receive the message - {e}')
+        except ValueError as e:
+            print(f'Value error, most probably while parsing MSG - msg: {msg}\n {e}')
+            return False
+        except Exception as e:
+            print(f'Exception while receiving a message - {e}')
+            return False
+
+        return True
+
+    @staticmethod
+    async def send_notification(notification: Notification):
+        """
+        Sends the following JSON to the recipient
+        {
+            "type": "NOTIFICATION",
+            "notification": {...}
+        }
+        """
+        recipient_id = notification.recipient_id
+        if recipient_id not in ws_connections or not ws_connections[recipient_id].is_valid:
+            print(f'Notification recipient with ID {recipient_id} was either not connected or not authorized')
+            return
+
+        asyncio.ensure_future(ws_connections[recipient_id].send_message({
+            "type": "NOTIFICATION",
+            "notification": NotificationSerializer(notification).data
+        }))
+
+
+async def authenticate_user(stream):
+    """
+    Authenticates the user, essentially validating his connection
+        and proving he is who he claims to be
+    Expects the following JSON
+    {
+    "type": "authentication",
+    "token": YOUR_NOTIFICATION_TOKEN_HERE,
+    "user_id": YOUR_ID_HERE,
+    }
+    Can either return
+        an error message
+        {
+            "type": "INVALID_NOTIFICATION_TOKEN",
+            "message": "Notification token is invalid or expired!"
+        }
+        an OK acknowledgement
+        {
+            "type": "OK",
+            "message": "Successfully authenticated"
+        }
+    """
+    while True:
+        auth_message = await stream.get()
+        token, user_id = auth_message.get('token'), auth_message.get('user_id')
+        if user_id not in ws_connections:
+            print(f'Somebody else tried to authenticate user_id {user_id}.')
+            continue
+
+        user_connection: UserConnection = ws_connections[user_id]
+        if not user_connection.user.notification_token_is_valid(token):
+            # TODO: This means that User 2 can spam an invalid token posing as User 3 and User 3 will receive a series of ERROR messages
```

Comment:
What is the severity of this?

Reply:
Pretty big, it meant the ability to spam a user's front-end client with countless error messages. *fun fact: if the front-end was coded to always send a request to the backend for a new notification token once it received a `NOTIFICATION_TOKEN_EXPIRED` message from the notifications websocket, this would allow for some indirect DDOS attacks, where a hacker would simply spam the `notifications` server with fake notification tokens, masking himself as user 1. User 1 would receive messages through the websocket that his token was expired and his front-end client would send a request to the back-end to get a new token.*

**I fixed it, though** - it was a fundamental problem in the server architecture. Due to me not thinking things through, we previously needed to authorize **every** websocket message being sent from the front-end by requiring the front-end to send a token tied to the user he claims to be. I remade it so the back-end remember which websocket connection is which user and those tokens can now be removed (maybe when we get to implementing the stuff)